crates/trie/trie/src/proof_v2/mod.rs

```diff
@@ -570,27 +646,454 @@ where
+                    self.cached_branch_stack.push((cached_path, cached_branch));
+
+                    continue
+                }
+            }
+
+            // We now want to check if there is a cached branch node at this child. The cached
+            // branch node may be the node at this child directly, or this child may be an
+            // extension and the cached branch is the child of that extension.
+
+            // All trie nodes prior to `child_path` will not be modified further, so we can seek the
+            // trie cursor to the next cached node at-or-after `child_path`.
+            if trie_cursor_state.path().is_some_and(|path| path < &child_path) {
+                trace!(target: TRACE_TARGET, ?child_path, "Seeking trie cursor to child path");
+                *trie_cursor_state = TrieCursorState::new(self.trie_cursor.seek(child_path)?);
+            }
+
+            // If the next cached branch node is a child of `child_path` then we can assume it is
+            // the cached branch for this child. We push it onto the `cached_branch_stack` and loop
+            // back to the top.
+            if let TrieCursorState::Available(next_cached_path, next_cached_branch) =
+                &trie_cursor_state &&
+                next_cached_path.starts_with(&child_path)
+            {
+                // Push the current cached branch back on before pushing its child and then looping
+                self.cached_branch_stack.push((cached_path, cached_branch));
+
+                trace!(
+                    target: TRACE_TARGET,
+                    ?child_path,
+                    ?next_cached_path,
+                    ?next_cached_branch,
+                    "Pushing cached branch for child",
+                );
+                self.cached_branch_stack.push(trie_cursor_state.take());
+                continue;
+            }
+
+            // There is no cached data for the sub-trie at this child, we must recalculate the
+            // sub-trie root (this child) using the leaves. Return the range of keys based on the
+            // child path.
+            let child_path_upper = child_path.increment();
+            trace!(
+                target: TRACE_TARGET,
+                lower=?child_path,
+                upper=?child_path_upper,
+                "Returning sub-trie's key range to calculate",
+            );
+
+            // Push the current cached branch back onto the stack before returning.
+            self.cached_branch_stack.push((cached_path, cached_branch));
+
+            return Ok(Some((child_path, child_path_upper)));
+        }
+    }
+
     /// Internal implementation of proof calculation. Assumes both cursors have already been reset.
     /// See docs on [`Self::proof`] for expected behavior.
     fn proof_inner(
         &mut self,
         value_encoder: &VE,
-        targets: impl IntoIterator<Item = Nibbles>,
+        targets: impl IntoIterator<Item = B256>,
```

Comment:
The changes to `should_retain` necessitated that targets are passed in as B256, so that they are in both lexicographic and depth-first order. This should be fine, I can't think of any use-cases where we'd want to pass in paths which aren't full keys anyway.